apps/total-typescript/src/pages/tutorials/[module]/completed.tsx

```diff
@@ -0,0 +1,322 @@
+import React from 'react'
+import Layout from 'components/app/layout'
+import Spinner from 'components/spinner'
+import queryString from 'query-string'
+import toast from 'react-hot-toast'
+import Image from 'next/image'
+import cx from 'classnames'
+import z from 'zod'
+import {convertToSerializeForNextResponse} from '@skillrecordings/commerce-server'
+import {getSdk, User} from '@skillrecordings/database'
+import {zodResolver} from '@hookform/resolvers/zod'
+import {SanityDocument} from '@sanity/client'
+import {Subscriber} from 'lib/convertkit'
+import {getModule} from 'lib/tutorials'
+import {GetServerSideProps} from 'next'
+import {get, isEmpty} from 'lodash'
+import {trpc} from 'utils/trpc'
+import {
+  FieldValues,
+  SubmitHandler,
+  useForm,
+  UseFormReturn,
+} from 'react-hook-form'
+import {
+  CopyToClipboard,
+  Facebook,
+  LinkedIn,
+  Twitter,
+} from '@skillrecordings/react'
+import {track} from 'utils/analytics'
+import {USER_ID_QUERY_PARAM_KEY} from '.'
+
+export const getServerSideProps: GetServerSideProps = async ({
+  res,
+  params,
+  req,
+}) => {
+  const tutorial = await getModule(params?.module as string)
+  let subscriber = req.cookies['ck_subscriber'] || null
+  subscriber = subscriber && JSON.parse(subscriber)
+
+  const {getUserByEmail} = getSdk()
+  const user =
+    subscriber && (await getUserByEmail(get(subscriber, 'email_address')))
+
+  if (!tutorial) {
+    return {
+      notFound: true,
+    }
+  }
+
+  if (!user) {
+    return {
+      redirect: {
+        destination: `/tutorials/${tutorial.slug.current}`,
+        permanent: false,
+      },
+    }
+  }
+
+  res.setHeader('Cache-Control', 's-maxage=1, stale-while-revalidate')
+
+  return {
+    props: {
+      tutorial,
+      subscriber,
+      user: convertToSerializeForNextResponse(user),
+    },
+  }
+}
+
+type CompletedTutorialProps = {
+  tutorial: SanityDocument
+  subscriber: Subscriber
+  user: User
+}
+
+const CompletedTutorialPage: React.FC<CompletedTutorialProps> = ({
```

Comment:
we currently aren't checking whether they've actually completed the tutorial so anyone can view this page